---
title: "DataType-Modality Representation"
author: "Dario Righelli"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
    BiocStyle::html_document:
      toc_float: true
vignette: >
    %\VignetteIndexEntry{DataType-Modality Representation}
    %\VignetteEncoding{UTF-8}
    %\VignetteEngine{knitr::rmarkdown}
Package: SingleCellMultiModal
bibliography: ../inst/REFERENCES.bib
editor_options:
  chunk_output_type: console
---

# Installation 

```{r,eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("SingleCellMultiModal")
```


# Load libraries

```{r, include=TRUE, results="hide", message=FALSE, warning=FALSE}
library(MultiAssayExperiment)
library(SingleCellMultiModal)
library(ggplot2)
```

# Loading datasets

This step will download all the datasets present into the package, it may take 
a while if you don't have any already cached on your local machine.

```{r}

mae <- SingleCellMultiModal(c("mouse_gastrulation", "pbmc_10x", "cord_blood", "peripheral_blood", "mouse_embryo_8_cell", "macrophage_differentiation", "mouse_visual_cortex"), versions=c("2.0.0", "1.0.0", "1.0.0","1.0.0","1.0.0","1.0.0","2.0.0"), dry.run=FALSE)

```

# Creating data structure

The following (not so elegant) code will create a `data.frame` used for the plot

```{r}

dimexp <- lapply(experiments(mae), dim)

dimslist <- list("CITEseq"=dimexp[12:13])
dimslist <- c(dimslist, list("ECCITEseq"=dimexp[14:16]))
dimslist <- c(dimslist, list("SCoPE2"=dimexp[19:20]))
dimslist <- c(dimslist, list("scNMT"=dimexp[1:9]))
dimslist <- c(dimslist, list("TENxMultiome"=dimexp[10:11]))
dimslist <- c(dimslist, list("seqFISH"=dimexp[21:22]))
dimslist <- c(dimslist, list("GTseq"=dimexp[17:18]))

dimensions <- data.frame()
k=1
for(i in 1:length(dimslist))
{
    if ( !isEmpty(grep("scNMT", names(dimslist)[i])) )
    {
        idx <- grep("acc", names(dimslist[[i]]))
        feat <- sum(unlist(lapply(dimslist[[i]][idx], function(x) return(x[1]))))
        cells <- median(unlist(lapply(dimslist[[i]][idx], function(x) return(x[2]))))
        dimensions[k,1] <- names(dimslist)[i]
        dimensions[k,2] <- "ATAC"
        dimensions[k,3] <- feat
        dimensions[k,4] <- cells
        k <- k+1
        idx <- grep("met", names(dimslist[[i]]))
        feat <- sum(unlist(lapply(dimslist[[i]][idx], function(x) return(x[1]))))
        cells <- median(unlist(lapply(dimslist[[i]][idx], function(x) return(x[2]))))
        dimensions[k,1] <- names(dimslist)[i]
        dimensions[k,2] <- "Methylation"
        dimensions[k,3] <- feat
        dimensions[k,4] <- dimslist[[i]][[1]][2]
        k <- k+1
        idx <- grep("rna", names(dimslist[[i]]))
        dimensions[k,1] <- names(dimslist)[i]
        dimensions[k,2] <- "RNA"
        dimensions[k,3] <- dimslist[[i]][[idx]][1]
        dimensions[k,4] <- dimslist[[i]][[idx]][2]
        k <- k+1
    } 
    
    for(j in 1:length(dimslist[[i]]))
    {
        if ( !isEmpty(grep("scNMT", names(dimslist)[i])) ){ next }
        dimensions[k,1] <- names(dimslist)[i]
        dimensions[k,2] <- names(dimslist[[i]])[j]
        dimensions[k,3] <- dimslist[[i]][[j]][1]
        dimensions[k,4] <- dimslist[[i]][[j]][2]
        k <- k+1
    }
}
colnames(dimensions) <- c("DataType", "Modality", "Features", "Cells")

dimensions$Modality[grep("seqFISH", dimensions$Modality)] <- "Spatial"

dimensions$Modality[grep("rna", dimensions$Modality)] <- "RNA"
dimensions$Modality[grep("RNA", dimensions$Modality)] <- "RNA"

dimensions$Modality[grep("protein", dimensions$Modality)] <- "Protein"
dimensions$Modality[grep("atac", dimensions$Modality)] <- "ATAC"

dimensions$Modality[grep("scADT", dimensions$Modality)] <- "ADT"
dimensions$Modality[grep("scHTO", dimensions$Modality)] <- "HTO"

dimensions$Modality[grep("genomic", dimensions$Modality)] <- "DNA"
dimensions$Modality[grep("transcriptomic", dimensions$Modality)] <- "RNA"

dimensions$Category <- NA
dimensions$Category[dimensions$Modality=="ADT"] <- "Proteomics"
dimensions$Category[dimensions$Modality=="HTO"] <- "Proteomics"
dimensions$Category[dimensions$Modality=="Protein"] <- "Proteomics"

dimensions$Category[dimensions$Modality=="RNA"] <- "Transcriptomics"

dimensions$Category[dimensions$Modality=="ATAC"] <- "Genomics"
dimensions$Category[dimensions$Modality=="Methylation"] <- "Genomics"
dimensions$Category[dimensions$Modality=="DNA"] <- "Genomics"

dimensions$Category[dimensions$Modality=="Spatial"] <- "Spatial"

dimensions <- dimensions[c(15, 16, 8:10, 11:12, 1:5, 6:7, 13:14),]
dimensions$DataType <- factor(dimensions$DataType, levels=unique(dimensions$DataType))

```


# DataType vs Modality 

The following plot shows on x-axis the datasets while on the y-axis the available
modalities.
The color code and the dot radius respectively represent the number of features 
and the number of cells per each dataset-modality combination.

```{r}

ggplot(dimensions) + 
    geom_point(aes(x=DataType, y=Modality, size=Cells, color=Features)) + 
    scale_size_continuous(range=c(3, 10)) +
    # scale_color_continuous() +
    scale_color_gradient(low="#56B4E9", high="#F0E442") +
    facet_wrap( ~ Category, ncol=1, scales="free_y") + 
    theme_bw()

```

